x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/action/PutSamlServiceProviderRequest.java

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+package org.elasticsearch.xpack.idp.action;
+
+import org.elasticsearch.ElasticsearchParseException;
+import org.elasticsearch.action.ActionRequest;
+import org.elasticsearch.action.ActionRequestValidationException;
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.ValidationException;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.xpack.idp.saml.sp.SamlServiceProviderDocument;
+
+import java.io.IOException;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.Objects;
+
+import static org.elasticsearch.action.ValidateActions.addValidationError;
+
+public class PutSamlServiceProviderRequest extends ActionRequest {
+
+    private final SamlServiceProviderDocument document;
+
+    public static PutSamlServiceProviderRequest fromXContent(String entityId, XContentParser parser) throws IOException {
+        final SamlServiceProviderDocument document = SamlServiceProviderDocument.fromXContent(null, parser);
+        if (document.entityId == null) {
+            document.setEntityId(entityId);
+        } else if (entityId != null) {
+            if (entityId.equals(document.entityId) == false) {
+                throw new ElasticsearchParseException(
+                    "Entity id [{}] inside request body and entity id [{}] from parameter do not match", document.entityId, entityId);
+            }
+        }
+        if (document.created != null) {
+            throw new ElasticsearchParseException(
+                "Field [{}] may not be specified in a request", SamlServiceProviderDocument.Fields.CREATED_DATE);
+        }
+        if (document.lastModified != null) {
+            throw new ElasticsearchParseException(
+                "Field [{}] may not be specified in a request", SamlServiceProviderDocument.Fields.LAST_MODIFIED);
+        }
+        document.setCreatedMillis(System.currentTimeMillis());
+        document.setLastModifiedMillis(System.currentTimeMillis());
+        return new PutSamlServiceProviderRequest(document);
+    }
+
+    public PutSamlServiceProviderRequest(SamlServiceProviderDocument document) {
+        this.document = document;
+    }
+
+    public PutSamlServiceProviderRequest(StreamInput in) throws IOException {
+        this.document = new SamlServiceProviderDocument(in);
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        document.writeTo(out);
+    }
+
+    public SamlServiceProviderDocument getDocument() {
+        return document;
+    }
+
+    @Override
+    public ActionRequestValidationException validate() {
```

Comment:
The `fromXContent` split was my first consideration as well, but it's hard to do nicely with `ObjectParser`. It's _possible_, but none of the options felt like they were an overall improvement.

We can pass a "mode" (probably just `Boolean`) as the "context" to a parser, but we already use the context, so we'd need to pass some sort of tuple of `SamlServiceProviderDocument` and `Boolean`.

Or (and this was my preferred option) we could have 2 parsers, one of which registers the right fields for the _Document_ and another registers the fields for the _Request_.
That's not too bad, but it kind of pushes the validation logic into the parser building (indirectly) so the logic is still split & it's harder to follow (because you need to understand that there are 2 parsers with different fields to enforce that the request doesn't permit certain fields from being set).

To be honest, we always do _some_ validation during XContent parsing. At the very least it's enforcing field names/types, but often it's more than that. This code does that more obviously than most, so it feels extra strange.

To my thinking, the root of the issue is that the `Request` lifecycle expects validation to happen separately to parsing, and that can be an artificial separation. It's a by-product of transport client (because you could have Requests that didn't come from XContent), but it's sometimes a weird fit.

I could approach this differently, and embrace that separation - which would mean if a field is not supposed to be set on the request, it should be `null` during `validate`. That is, don't treat it as a _parsing_ issue, but as a request issue.
Then I'd need to add some conditional logic in doc.validate() (or maybe duplicate the logic into the request) so that fields which are supposed to be null in the request but cannot be null in the index are validated correctly.
  
I think that would work, but it has the slightly weird behaviour that we accept certain fields during parsing that really shouldn't be part of the API (but would be validated when the request is processed).

Reply:
Let's chat. I think we can do better than what I've put here, but it might be worth making it a new PR so that we can debate possible options.

Reply:
We discussed this, and I will raise a separate PR for it so that we can move forward with this change, and then iterate on that.